### PR TITLE
Dev/firestore refactor

### PR DESCRIPTION
**Pull request by TheBlueHawk**
Welcome to FirestoreInteractor 3.0! 
No more wrapper, no more additional Interactors, only a centralized parameter-less ConcreteFirestoneInteractor which support several flexible queries: read-document/read full collection, write a doc/write with a specific ID, use callbacks or QueryHandler/onXXXXListener

### CODE REVIEW

**Review by TheCrazyKing — APPROVED** [review]
Nice Firebase 3.0 !! 
I see that you changed the min SDK to 24 as we discussed on the last meeting. 
Great.

File: app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import com.google.firebase.firestore.QuerySnapshot;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import ch.epfl.sdp.Callback;
+
+public abstract class FirestoreInteractor {
```
[the remark this answers is not in the document]
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java:20]: The method surcharge is made on purpose to offer more flexibility when interacting with Firestore

File: app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        return task -> {
+            if (task.isSuccessful()) {
+                handler.onSuccess(task.getResult());
+            } else {
+                handler.onFailure();
+            }
+        };
+    }
+
+    OnCompleteListener<DocumentSnapshot> documentSnapshotOnCompleteListener(Callback callback) {
```
[the remark this answers is not in the document]
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java]: Dead code, will be removed with the next commit

File: app/src/main/java/ch/epfl/sdp/firestore/ConcreteFirestoreInteractor.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            collectionReference
+                    .add(document)
+                    .addOnSuccessListener(successListener)
+                    .addOnFailureListener(failureListener);
+        } finally {
+            serverIdlingResource.decrement();
+        }
+    }
+
+    public void writeDocumentWithID(DocumentReference documentReference, Object document,
```
[the remark this answers is not in the document]
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/firestore/ConcreteFirestoreInteractor.java:30]: Similar blocks but they use different functions on different objects

File: app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/firestore/FirestoreInteractor.java:20]: The method surcharge is made on purpose for increased flexibility when interacting with Firestore

**Review by lucieho — APPROVED** [review]
Great, looks much more handy!

**Pull request merged**